mirgecom/profiling.py

```diff
@@ -0,0 +1,207 @@
+    time: int
+    flops: int
+    bytes_accessed: int
+    footprint_bytes: int
+
+
+@dataclass
+class ProfileEvent:
+    """Class to hold a profile event that has not been seen by the profiler yet."""
+    cl_event: cl._cl.Event
+    program: lp.kernel.LoopKernel
+    kwargs: dict
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: tabulate_profiling_data
+    .. automethod:: call_loopy
+
+    Inherits from :class:`meshmode.array_context.PyOpenCLArrayContext`."""
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue. "
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def _finish_profile_events(self) -> None:
+        # First, wait for event completion
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        for t in self.profile_events:
+            program = t.program
+            res = self._get_kernel_stats(program, t.kwargs)
+            res.time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            self.profile_results.setdefault(program, []).append(res)
+
+        self.profile_events = []
+
+    def tabulate_profiling_data(self) -> pytools.Table:
+        """Returns a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls", "Time_min [s]", "Time_avg [s]",
+            "Time_max [s]", "GFlops_min", "GFlops_avg", "GFlops_max", "GMemAcc_min",
+            "GMemAcc_avg", "GMemAcc_max", "Bandwidth [GByte/s]",
+            "Footprint [GByte]"])
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v.time / 1e9 for v in value]
+            flops = [v.flops / 1e9 for v in value]
+            bytes_accessed = [v.bytes_accessed / 1e9 for v in value]
+            footprint_bytes = [v.footprint_bytes / 1e9
+                if v.footprint_bytes is not None else 0 for v in value]
+
+            g = ".4g"
+
+            tbl.add_row([key.name, num_values, f"{min(times):{g}}",
+                f"{mean(times):{g}}", f"{max(times):{g}}", f"{min(flops):{g}}",
+                f"{mean(flops):{g}}", f"{max(flops):{g}}",
+                f"{min(bytes_accessed):{g}}", f"{mean(bytes_accessed):{g}}",
+                f"{max(bytes_accessed):{g}}",
+                f"{round(mean(bytes_accessed)/mean(times), 3):{g}}",
+                f"{mean(footprint_bytes):{g}}"])
```

Comment:
Absolute gflops/gbyte numbers tell me very little. I would rather see:

* flop rate (min/max/mean)
* bandwidth (according to access count, min/max/mean)
* bandwidth (according to footprint, min/max/mean)

Reply:
Added the values you indicated, and removed the raw flops, bytes numbers.